### PR TITLE
fix RegionalPartner.find_by_region method to handle null inputs

### PR DESCRIPTION
**Pull request by sureshc**
https://trello.com/c/GRNOraZ8/274-add-tests-for-regionalpartnerfindbyregion-and-make-sure-it-handles-null-inputs

Also add tests for this method.

### CODE REVIEW

**Comment by sureshc** [comment]
[Add tests for RegionalPartner.find_by_region, and make sure it handles null inputs](https://trello.com/c/GRNOraZ8/274-add-tests-for-regionalpartnerfindbyregion-and-make-sure-it-handles-null-inputs)

File: dashboard/app/models/regional_partner.rb [hunk 1]
```diff
@@ -58,16 +58,24 @@ def program_manager=(program_manager_id)
...
+      if zip_code
+        find_by_region_query = find_by_region_query.or(
+          joins(:mappings).
+          where(pd_regional_partner_mappings: {zip_code: zip_code})
+        )
+      end
+    elsif zip_code
+      find_by_region_query = find_by_region_query.
+        joins(:mappings).where(pd_regional_partner_mappings: {zip_code: zip_code})
+    end
```
Comment (aoby) [inline dashboard/app/models/regional_partner.rb]: This looks a bit cumbersome. The `joins(:mappings)` is repeated in all cases, and the zip_code query is repeated. 

How about something like this instead?
```ruby
base_query = RegionalPartner.joins(:mappings)
queries = []
queries << base_query.where(pd_regional_partner_mappings: {state: state}) if state
queries << base_query.where(pd_regional_partner_mappings: {zip_code: zip_code}) if zip_code

full_query = queries.length == 1 ? queries[0] : queries[0].or(queries[1])
```

File: dashboard/test/models/regional_partner_test.rb [hunk 2]
```diff
@@ -65,4 +65,59 @@ class RegionalPartnerTest < ActiveSupport::TestCase
...
     end
   end
+
+  test 'find_by_region finds matching regional_partner' do
+    # partners and their mappings
+
+    # regional_partner_WA
+    # regional_partner_WA_98105
+    # regional_partner_98105
+    # regional_partner_NY
```
Comment (aoby) [inline dashboard/test/models/regional_partner_test.rb]: What are these comments?

File: dashboard/test/models/regional_partner_test.rb [hunk 3]
```diff
@@ -65,4 +65,59 @@ class RegionalPartnerTest < ActiveSupport::TestCase
...
+    # regional_partner_CA_94305
+    regional_partner_ca_94305 = create :regional_partner, name: "partner_CA_94305"
+    regional_partner_ca_94305.mappings.find_or_create_by(state: "CA")
+    regional_partner_ca_94305.mappings.find_or_create_by(zip_code: "94305")
+
+    # regional_partner_70808
+    regional_partner_70808 = create :regional_partner, name: "partner_70808"
+    regional_partner_70808.mappings.find_or_create_by(zip_code: "70808")
+
+    # regional_partner_no_mappings
```
Comment (aoby) [inline dashboard/test/models/regional_partner_test.rb]: This too. The test seems incomplete, or the comments are extra. Also I don't think the comments above for each  variable declaration add any value.

File: dashboard/test/models/regional_partner_test.rb [hunk 4]
```diff
@@ -65,4 +65,59 @@ class RegionalPartnerTest < ActiveSupport::TestCase
...
+
+    # queries
+
+    # state=nil/zip=nil [return nil (no match)]
+    assert_nil RegionalPartner.find_by_region(nil, nil)
+
+    # state=NY [search by state only where there is only one match]
+    assert_equal RegionalPartner.find_by_region(nil, "NY"), regional_partner_ny
```
Comment (aoby) [inline dashboard/test/models/regional_partner_test.rb]: Reverse the args. [assert_equal](http://ruby-doc.org/stdlib-2.1.0/libdoc/test/unit/rdoc/Test/Unit/Assertions.html#method-i-assert_equal) takes expected followed by actual.

Same below.

File: dashboard/test/models/regional_partner_test.rb [hunk 5]
```diff
@@ -65,4 +65,59 @@ class RegionalPartnerTest < ActiveSupport::TestCase
...
+    # zip=90210 [zip only, no match]
+    assert_nil RegionalPartner.find_by_region("90210", nil)
+
+    # state=CA/zip=90210 [state matches one partner, zip code does not match]
+    assert_equal RegionalPartner.find_by_region("94305", "CA"), regional_partner_ca_94305
+
+    # state=LA/zip=70808 [zip code matches one partner, state does not match]
+    # state=WA/zip=98104 [state matches many, zip matches one]
+    # state=WA [state matches many, indeterminate result]
+    # zip=98105 [zip matches many, indeterminate result]
```
Comment (aoby) [inline dashboard/test/models/regional_partner_test.rb]: missing coverage?

File: dashboard/test/models/regional_partner_test.rb (hunk 2 above)
Comment (sureshc) [reply dashboard/test/models/regional_partner_test.rb]: Sorry for not explaining - those comments were placeholders for test data scenarios that were not complete yet.  Most of the test scenarios are complete now.

File: dashboard/app/models/regional_partner.rb (hunk 1 above)
Comment (sureshc) [reply dashboard/app/models/regional_partner.rb]: Yup.  I found a variation on that approach that works.

**Comment by sureshc** [comment]
@aoby I'm having trouble with the "state matches many, zip matches one" test.  I've manually configured the same set of Regional Partners and Mappings in my `development` environment that are created in the tests and I can run all of the tests successfully in my dashboard console.  However, `RegionalPartner.find_by_region("98104", "WA")` in the `test` environment incorrectly returns the first Regional Partner in the database with a matching state instead of preferring the one with the matching zip_code as well.  This is the SQL the query method generates:

``` sql
SELECT  `regional_partners`.* 
FROM `regional_partners`
INNER JOIN `pd_regional_partner_mappings` ON `pd_regional_partner_mappings`.`regional_partner_id` = `regional_partners`.`id`
WHERE (`pd_regional_partner_mappings`.`state` = 'WA'
  OR `pd_regional_partner_mappings`.`zip_code` = '98104')
ORDER BY pd_regional_partner_mappings.zip_code IS NOT NULL DESC LIMIT 1
```

The `ORDER BY zip_code IS NOT NULL` is critical for ensuring that zip_code matches are preferred.  I can run this query directly against mysql and tinker with the ORDER BY clause to confirm it works.

**Pull request merged**